Fix minor issues with password-handling, refactor

nbraud opened the pull request:
## Bug fixes

- Use a pw showing Unicode-related issue in `tests/api/auth_login`
- Use unicode normalization, to fix that issue.
- Use constant-time equality to compare password hashes.  
  Regular string comparison takes a time that depends on where the hashs
  first differ, which can leak information about the password hash.


## Refactor

- Extract password handling from `app.models` to `app.pw_models`
- Refactor `tests/api/test_auth_login` to avoid copy-pasted code between tests
- Remove the separate `salt` column, `bcrypt` encodes it as part of the “hash”.  
  This will *not* break existing accounts.
  ~~I couldn't run `new_migration.sh` (Docker-related issues, apparently)~~


## Context

I ran into those issues while trying to port Simple Login to modern
password hashing practices.  I thought it best to make a first PRs
with all the simpler changes, while I keep working on this.

nguyenkims commented:
Hey thanks for making this PR! 

> Remove the separate salt column, bcrypt encodes it as part of the “hash”.

Does that mean the `User.password` already contain `User.salt` and therefore the `User.salt` column can be dropped? This is indeed very interesting and I didn't know that! Do you have a link or article writing about this :) ?

Btw I wonder if we can make the PR smaller by making the change directly in the current `check_password` and `set_password` method instead of creating `app.pw_models`? Adding `app.pw_models` seems to be rather a refactoring (a good one as User class is very long now!) but I think it's better to separate feature adding/modification PR to refactoring PR, what do you think?

nbraud commented:
> > Remove the separate salt column, bcrypt encodes it as part of the “hash”.
> 
> Does that mean the `User.password` already contain `User.salt` and therefore the `User.salt` column can be dropped? This is indeed very interesting and I didn't know that! Do you have a link or article writing about this :) ?

Yes, exactly.  `bcrypt` itself doesn't document that explicitly, but:
- it provides a function `checkpw(password: bytes, hashed_password: bytes) -> bool`, which is only possible of the salt is part of the `hashed_password` argument;
- it implements a [glibc-style `crypt(3)`] hash format: `$2b$12$<salt><hash>`.

[glibc-style `crypt(3)`]: https://manpages.debian.org/buster/manpages-dev/crypt.3.en.html#Features_in_glibc

> Btw I wonder if we can make the PR smaller by making the change directly in the current `check_password` and `set_password` method instead of creating `app.pw_models`? Adding `app.pw_models` seems to be rather a refactoring (a good one as User class is very long now!) but I think it's better to separate feature adding/modification PR to refactoring PR, what do you think?

I included such a change because this PR is intended to be only refactoring and bugfixes, yes.
In addition, `models.py` is large-enough that working directly in there was much more stress-inducing.

nguyenkims commented:
@nbraud thanks for the info! As the salt is **implicitly** stored in the hashed password, I think it's better to keep storing the salt, what do you think?

For the PR, I wonder if a change like this can be enough? https://github.com/simple-login/app/commit/183cd8c6b79305080919c7a23802fa0612fd6820

I prefer to separate bug fixes and refactoring into different PR for safety reason :), especially when dealing with password like this one :).

nbraud commented:
> @nbraud thanks for the info! As the salt is **implicitly** stored in the hashed password, I think it's better to keep storing the salt, what do you think?

Why do you believe this is better?  Having duplicate data seems strictly worse to me, since it means the two columns can have inconsistent values for the salt.  (In database terms, the current schema isn't normalized.)

More importantly, I replaced the logic in `check_password` with a call to `bcrypt.checkpw` to fix the use of a non-constant-time equality check (which is problematic from a security standpoint).  After that change, nothing reads from the `salt` column anymore.

> For the PR, I wonder if a change like this can be enough? [183cd8c](https://github.com/simple-login/app/commit/183cd8c6b79305080919c7a23802fa0612fd6820)

That seems entirely inadequate: it only corrects one of the bugs I fixed (the Unicode one, which isn't security-relevant) and doesn't include tests for the change.

> I prefer to separate bug fixes and refactoring into different PR for safety reason :), especially when dealing with password like this one :).

I don't see the safety reasons here, but I could split the PR along those lines if you insist:
1. refactoring (no functional changes): commits 52d4d2a and f5f4d46 ;
2. fix the unicode bug, add a test for it: commits d216812 and ecd74b8 ;
3. fix the use of variable-time equality, drop the `salt` column: commit 586654e.

Note that (2) and (3) would depend on (1) (if cherry-picking the commits) so the refactoring would have to go in first.

nguyenkims commented:
> Why do you believe this is better? Having duplicate data seems strictly worse to me, since it means the two columns can have inconsistent values for the salt. (In database terms, the current schema isn't normalized.)

I agree that having duplicated data is a bad design. But as the "salt" isn't explicitly documented, we can consider bcrypt requires it (as other algorithms would). I also wonder if the salt can be useful in case we want to replace bcrypt by other algorithms like Argon2.

> I don't see the safety reasons here, but I could split the PR along those lines if you insist:

It'd be great if we can have small PR like you mentioned. I still insist on not removing "salt" column for now :).

TheLastProject commented:
> I agree that having duplicated data is a bad design. But as the "salt" isn't explicitly documented, we can consider bcrypt requires it (as other algorithms would).

Well, yes, bcrypt requires it. But bcrypt already stores it in the password column. See this example:

```
@>>> import bcrypt
@>>> salt = bcrypt.gensalt()
@>>> hash = bcrypt.hashpw("password".encode(), salt).decode()

@>>> print(salt.decode())
$2b$12$bD4Apn9MWeZICtfMHQuOU.
@>>> print(hash[:29])
$2b$12$bD4Apn9MWeZICtfMHQuOU.
@>>> salt.decode() == hash[:29]
True
```

Try it on your own database, you'll see that the salt is always the same as the first 29 characters of each password. So even if you somehow do need to extract that salt later, you can just grab the first 29 characters. No need to store it twice, really.

AstraLuma commented:
> I also wonder if the salt can be useful in case we want to replace bcrypt by other algorithms like Argon2.

On the subject of prior art, django's `django.contrib.auth` only has a password column, with the various knobs embedded in it.

Model: https://github.com/django/django/blob/e703b152c6148ddda1b072a4353e9a41dca87f90/django/contrib/auth/base_user.py#L47

Hashing & checking core: https://github.com/django/django/blob/main/django/contrib/auth/hashers.py

This is a configurable and pluggable password system, supporting bcrypt, argon2, PBKDF2, and lesser algorithms.

nbraud commented:
> > Why do you believe this is better? Having duplicate data seems strictly worse to me, since it means the two columns can have inconsistent values for the salt. (In database terms, the current schema isn't normalized.)
> 
> I agree that having duplicated data is a bad design. But as the "salt" isn't explicitly documented, we can consider bcrypt requires it (as other algorithms would).

1. As I already stated, `bcrypt`'s API does guarantee that we do not need a separate copy of the salt to verify a password, since there is a function `checkpw(password: bytes, hashed_password: bytes) -> bool`.

2. As I already stated, `bcrypt` produces an output in a format that's a documented standard; like with any other password hash using that standard, the salt and other parameters (here, the number of rounds) are encoded in the output.  
   This is nothing specific to `bcrypt`, any other password hash needs to provide something equivalent in order to be practical (since additional parameters must be encoded alongside the salt).

3. The **current** usage of `bcrypt` is incorrect in several ways:
   - There's no explicit API guarantee supporting the use of `hashpw(password, salt) == hashed_pw`.  
     The only provided way to verify a password, is the `checkpw` function, which does not takes a separate salt as input.
   - As I already stated, you are currently using variable-time string comparison, which is potentially vulnerable to [timing attacks].
[timing attacks]: https://en.wikipedia.org/wiki/Timing_attack

4. Regarding the desire for “explicit documentation”, I simply do not understand what you want:
   - I already [pointed you] to the relevant manpage
[pointed you]: https://github.com/simple-login/app/pull/496#issuecomment-851044088
   - `bcrypt` does explicitly expose a `checkpw` function; this isn't an internal function, or anything like that.
   - `bcrypt` **doesn't have any “explicit documentation”** beyond its function prototypes; this is literally all its module-level documentation:

```
NAME
    bcrypt

DESCRIPTION
    # Author:: Donald Stufft (<donald@stufft.io>)
    # Copyright:: Copyright (c) 2013 Donald Stufft
    # License:: Apache License, Version 2.0
    #
    # Licensed under the Apache License, Version 2.0 (the "License");
    # you may not use this file except in compliance with the License.
    # You may obtain a copy of the License at
    #
    # http://www.apache.org/licenses/LICENSE-2.0
    #
    # Unless required by applicable law or agreed to in writing, software
    # distributed under the License is distributed on an "AS IS" BASIS,
    # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    # See the License for the specific language governing permissions and
    # limitations under the License.

PACKAGE CONTENTS
    __about__
    _bcrypt

FUNCTIONS
    checkpw(password: bytes, hashed_password: bytes) -> bool
    
    gensalt(rounds: int = 12, prefix: bytes = b'2b') -> bytes
    
    hashpw(password: bytes, salt: bytes) -> bytes
    
    kdf(password: bytes, salt: bytes, desired_key_bytes: int, rounds: int, ignore_few_rounds: bool = False) -> bytes
```

@nguyenkims, I have to admit all of this makes me feel like you are using double standards: clearly, you didn't ask whoever wrote the current password-handling code to refer to “explicit documentation”, since it simply does not exist for Python's `bcrypt` implementation.

You have to understand that this is *incredibly exhausting* to deal with:
- I end up having to repeat myself again and again, which makes me feel like I'm not being listened to.
- You seem to be asking for some documentation that simply doesn't exist, so no matter what I do it's seemingly not good enough to get those issues fixed.


> I also wonder if the salt can be useful in case we want to replace bcrypt by other algorithms like Argon2.

No.

As I said initially, all those changes came up while trying to being Simple Login's password-handling implementation up to modern standards.

In particular, it's much easier between different password hashes (or more generally, password verifiers) with a single blob of data capturing all the data necessary to check a password — the hash itself, the salt, and all other data that's specific to a given password-hashing mechanism (for `bcrypt`, the number of rounds).

As @AstraLuma pointed out, it's a completely normal thing to do, and there's many projects doing exactly that.


> It'd be great if we can have small PR like you mentioned. 

OK, I'll do that.  I hope we can move forward more productively that way  :/

nguyenkims commented:
@nbraud  @TheLastProject sorry for the frustration! It seems that I have a wrong understanding about bcrypt and the current implementation reflects that!

After reading Django code (which is really well written, thanks @AstraLuma  !), everything is much clear now! Let me merge this PR, run it on our staging and run a round of smoke test.

Thank you all for the patience and sorry again for all the disappointment!
